Pass order currency to wc_price() when not in customer's session.

shendy-a8c opened the pull request:
Fixes #1200

#### Changes proposed in this Pull Request

* Passing order currency to `wc_price()`.
* Get intent currency from the response we get from the server.
* Save order intent currency as a fallback in case we cannot get the transaction currency from the server (for order note created from within catch block, for example).
* Handle zero decimal currency in failed refund webhook handler.

#### Testing instructions

1. Testing `capture_charge()`. Steps:
- Admin > Payments > Settings > Make sure `Issue an authorization on checkout, and capture later.` is checked.
- Install plugin that can change customer currency, like [WooCommerce Multi-Currency](https://woocommerce.com/products/multi-currency/).
- Setup currency plugin to the point where you can see non-USD currency from the front store, for example, for [WooCommerce Multi-Currency](https://woocommerce.com/products/multi-currency/), follow their [docs](https://docs.woocommerce.com/document/multi-currency) and open front store with parameter `?currency=EUR` to see prices in Euro (if you setup Euro as one of the customer currencies).
- Buy a product with certain credit card numbers listed below for each test case:
  - Success: 4242424242424242 with any expired date in the future and any 3-digit CVC code.
  - Error: (I cannot find suitable testing card number that can get us to the 2 possible non-success cases [[#](https://github.com/Automattic/woocommerce-payments/pull/1227/files#diff-bcfd303d507c33a27671935e724ee64d0df30ff2a18bb6ef4c94f28d48190124R1161)] [[#](https://github.com/Automattic/woocommerce-payments/pull/1227/files#diff-bcfd303d507c33a27671935e724ee64d0df30ff2a18bb6ef4c94f28d48190124R1172)], it has to be one that passes auth step but fails capture).
- Open the newly created order from admin.
- Select 'Capture charge' from the Order actions dropdown.
- Click the 'Update' button.
- With this branch, expect order note shows amount with correct order currency. With base branch, order note will use store currency.
- Test more cases where you change the order currency from the admin order page, for example [WooCommerce Multi-Currency](https://woocommerce.com/products/multi-currency/) can do that. Go to the very bottom of the order admin page and you will see a dropdown to change the order currency. Make sure to click the 'Update' button at the top right. After changing the order currency, go through the above steps again and make sure order note still uses the currency when the order was placed.

2. Testing `process_refund()`. Steps:
- Install plugin that can change customer currency, like [WooCommerce Multi-Currency](https://woocommerce.com/products/multi-currency/).
- Setup currency plugin to the point where you can see non-USD currency from the front store, for example, for [WooCommerce Multi-Currency](https://woocommerce.com/products/multi-currency/), follow their [docs](https://docs.woocommerce.com/document/multi-currency) and open front store with parameter `?currency=EUR` to see prices in Euro (if you setup Euro as one of the customer currencies).
- Buy a product with certain credit card numbers listed below for each test case:
  - Success refund: 4242424242424242 with any expired date in the future and any 3-digit CVC code.
  - Failed refund: 4000000000000259 with any expired date in the future and any 3-digit CVC code. This is to test the failed case. You will see an empty alert (filed already as [separate issue](https://github.com/Automattic/woocommerce-payments/issues/1203)). Refresh the page to see the order note. That should be to test the [failed case](https://github.com/Automattic/woocommerce-payments/pull/1227/files#diff-bcfd303d507c33a27671935e724ee64d0df30ff2a18bb6ef4c94f28d48190124R789).
- Open the newly created order from admin.
- Issue a refund, without and with reason, to test success cases respectively [[#](https://github.com/Automattic/woocommerce-payments/pull/1227/files#diff-bcfd303d507c33a27671935e724ee64d0df30ff2a18bb6ef4c94f28d48190124R804)] [[#](https://github.com/Automattic/woocommerce-payments/pull/1227/files#diff-bcfd303d507c33a27671935e724ee64d0df30ff2a18bb6ef4c94f28d48190124R810)].
- With this branch, expect order note shows amount with correct order currency. With base branch, order note will use store currency.
- Test more cases where you change the order currency from the admin order page, for example [WooCommerce Multi-Currency](https://woocommerce.com/products/multi-currency/) can do that. Go to the very bottom of the order admin page and you will see a dropdown to change the order currency. Make sure to click the 'Update' button at the top right. After changing the order currency, go through the above steps again and make sure order note still uses the currency when the order was placed.

3. Testing [refund failure webhook handler](https://github.com/Automattic/woocommerce-payments/pull/1227/files#diff-cea8dab22a014dc0e668f8fd148aaeecd2863b3eb09cbe1f9408092c9b8d6391R194) case.
- Purchase with testing card 4000000000005126 to make refund fail.
- Test with zero decimal currency like JPY to make sure amount is correct.
<img width="289" alt="Screen Shot 2021-02-01 at 09 19 53" src="https://user-images.githubusercontent.com/73803630/106407498-c937f580-646e-11eb-82fd-6ced8ca8fdfd.png">

4. Testing [subscription renewal hook](https://github.com/Automattic/woocommerce-payments/pull/1227/files#diff-a5e8d94b5af9b7536f23cf436b4eef1fb76da3069f34a63b4cf35c46051e417bR155).
I am not sure how to set this up. I just tested it with [unit test](https://github.com/Automattic/woocommerce-payments/pull/1227/files#diff-933977839629fb49196cf683653de08b29c2f319c0cc0edb91ecac15d1e747a6R256-R282).

-------------------

- [x] Added changelog entry (or does not apply)
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

<!--
Please read P7bbVw-3ww-p2 before opening the PR, assign the correct status to it, __and make sure that the related issue uses the Has PR status!__.
-->

shendy-a8c commented:
Questions:
1. Do I need to add changelog for this PR?
2. How to setup webhook on wcpay client? Won't that require a setup in Stripe for each self-hosted site? I thought only wcpay server that can accept webhook.

vbelolapotkov commented on includes/class-wc-payments.php:
```diff
@@ -146,6 +146,22 @@ public static function init() {
+	 * @param WC_Order $order The order whose original currency we want to save.
+	 */
+	public static function maybe_set_original_order_currency( $order ) {
+		$currency          = $order->get_currency();
+		$key               = '_wcpay_original_order_currency';
+		$original_currency = $order->get_meta( $key );
+		if ( ! $currency || $original_currency ) {
+			return;
+		}
```
Looks like a good place to reuse our own utility method `get_order_original_currency`?

kalessil replied:
@shendy-a8c : the order has `_order_currency` meta already, isn't it the same as we are trying to do here?

shendy-a8c replied:
@kalessil per [my comment here](https://github.com/Automattic/woocommerce-payments/issues/1200#issuecomment-766176096), Multi-currency app can change order's `_order_currency` and that can mess things up, so I save the original order currency in `_wcpay_original_order_currency`.

While we're talking about `_order_currency`, I'm actually confused with the difference between `$order->get_currency()` and `$order->get_meta('_order_currency')`. They seem to return the same value.

shendy-a8c replied:
@vbelolapotkov do you mean change that function block to be like this?
```
	public static function maybe_set_original_order_currency( $order ) {
		$currency          = $order->get_currency();
		$original_currency = WC_Payments_Utils::get_order_original_currency( $order );
		if ( ! $currency || $original_currency ) {
			return;
		}
		$order->update_meta_data( $key, $currency );
	}
```
or
```
	public static function maybe_set_original_order_currency( $order ) {
		$original_currency = WC_Payments_Utils::get_order_original_currency( $order );
		if ( ! $original_currency ) {
			return;
		}
		$order->update_meta_data( $key, $original_currency );
	}
```

If the latter, I prefer not because I like to be conservative and avoiding writing meta data when not necessary. It's safer that way, right?

kalessil replied:
@shendy-a8c : that's pretty strange, to reproduce I need to try refunding a non-usd order. Is that right?

`get_currrency` is a public API, which preferred way to access the currency (underlying data structures and fallback logic can evolve and the app still works).
`get_meta` is a lower-level API, any changes in underlying data structures can break app.

shendy-a8c replied:
> to reproduce I need to try refunding a non-usd order. Is that right?
Yes.
- Install [Multi-currency app](https://woocommerce.com/products/multi-currency/) and set it up, so you buy product with non-USD.
- Purchase something in non-USD, say in EUR.
- Open your order from admin.
- At the very bottom of the order admin page, you will see a dropdown to change the currency of the order. Checkout [my screenshot](https://user-images.githubusercontent.com/73803630/105613628-466fc480-5df6-11eb-970a-bd7ca3d47a0b.png). Change the order's currency to something else, say JPY.
- Refund the order.

Since refunding is not in customer's session, if we only pass `$order->get_currency()` to `wc_price()`, it will use JPY. That's not correct because the order was created in EUR. That's why I need to store 'original order currency' in `_wcpay_original_order_currency`, can't rely on `_order_currency` or `get_currency()`.

vbelolapotkov replied:
> If the latter, I prefer not because I like to be conservative and avoiding writing meta data when not necessary. It's safer that way, right?
The latter will call the update_meta_data() even though _wcpay_original_order_currency is already set.

I don't see it as a big problem. It's not a bottleneck to optimize it, but using internal meta key as a string literal bothers me more. How about moving the whole method (it's static after all) to the WC_Payments_Utils and moving internal meta key to private static property on Utils? This would make both usages next to each other and avoid using string literals for meta key.

shendy-a8c replied:
That's a fair concern, @vbelolapotkov. I've implemented your suggestion, ie moving the meta key to a constant in utils and put the set and get method close to each other within utils class, avoiding using string literal.

vbelolapotkov commented on tests/unit/admin/test-class-wc-rest-payments-webhook.php:
```diff
@@ -173,15 +173,21 @@ public function test_valid_failed_refund_update_webhook() {
 			->with(
 				$this->matchesRegularExpression(
-					'~^A refund of <span class="woocommerce-Price-amount amount">(<bdi>)?<span class="woocommerce-Price-currencySymbol">&pound;</span>9.99(</bdi>)?</span> was <strong>unsuccessful</strong> using WooCommerce Payments \(<code>test_refund_id</code>\).$~'
+					'~^A refund of <span class="woocommerce-Price-amount amount">(<bdi>)?<span class="woocommerce-Price-currencySymbol">&euro;</span>9.99(</bdi>)?</span> was <strong>unsuccessful</strong> using WooCommerce Payments \(<code>test_refund_id</code>\).$~'
```
🤔 Let's create a new similar test case for that. This applies to all other edited tests as well.

shendy-a8c replied:
Good point.

shendy-a8c replied:
Done.

vbelolapotkov reviewed (changes requested):
Hey @shendy-a8c! I've made some tests and it works very well 🎉 Except cases you've already mentioned where buttons and UI for refund shows other currency. 

Regarding the code, I found with global search a few cases where `wc_price` is used without currency, are they omitted intentionally?

Other general point, how about extracting the whole function into our own utility, e.g. `WC_Payments_Utils::format_order_currency( number $amount, WC_Order $order )` and reuse it in other places?

kalessil commented on includes/compat/subscriptions/class-wc-payment-gateway-wcpay-subscriptions-compat.php:
```diff
@@ -152,7 +152,7 @@ public function scheduled_subscription_payment( $amount, $renewal_order ) {
 						]
 					),
-					wc_price( $amount ),
+					wc_price( $amount, [ 'currency' => WC_Payments_Utils::get_order_original_currency( $renewal_order ) ] ),
```
With reference to my another comment, we could use `$renewal_order->get_currency()` here (and similar calls in refund processing).

shendy-a8c replied:
Answered here https://github.com/Automattic/woocommerce-payments/pull/1227#discussion_r564995909.

kalessil commented:
> Do I need to add changelog for this PR?
Yes, as this fixes refund currency in order notes with non-usd orders.

shendy-a8c commented:
@vbelolapotkov 
> Regarding the code, I found with global search a few cases where wc_price is used without currency, are they omitted intentionally?

I had that suspicion as well whether I need to pass currency to all `wc_price()` calls, but I think that's the importance of 'outside customer session' from the issue's title because I see that the currency changer app (Multi-currency in my case) hooks to `woocommerce_currency`.
The flow is like this:
- `wc_price()` is called.
- `get_woocommerce_currency_symbol()` is called [[#](https://github.com/woocommerce/woocommerce/blob/master/includes/wc-formatting-functions.php#L609)].
- `get_woocommerce_currency ()` is called [[#](https://github.com/woocommerce/woocommerce/blob/master/includes/wc-core-functions.php#L847)].
- filtered `get_option( 'woocommerce_currency' )` is returned [[#](https://github.com/woocommerce/woocommerce/blob/master/includes/wc-core-functions.php#L465)].

So most `wc_price()` already behaves correctly. We only need to take care when `wc_price()` is called outside customer's session.

kalessil commented on includes/admin/class-wc-rest-payments-webhook-controller.php:
```diff
@@ -190,7 +190,7 @@ private function process_webhook_refund_updated( $event_body ) {
 				]
 			),
-			wc_price( $amount / 100 ),
+			wc_price( $amount / 100, [ 'currency' => WC_Payments_Utils::get_order_original_currency( $order ) ] ),
```
In this case, we can access the currency information in the event itself: https://stripe.com/docs/api/refunds/object - performance, and context friendly.

shendy-a8c replied:
Good idea! Here it is [abb8804](https://github.com/Automattic/woocommerce-payments/pull/1227/commits/abb8804021f924f8a7488f5d13cd5bfd3fde1b65).

shendy-a8c commented:
- [x] Rebased.
- [x] Resolved conflict.

shendy-a8c commented:
- [x] Rebased.
- [x] Resolved conflict.
- [x] Add a placeholder for changelog that will give a conflict after the upcoming release.

shendy-a8c commented:
After successfully setting up webhook and testing failed refund case for real, I notice there is still bugs around:
- Rounding up amount, and
- Amount in cent.

<img width="1589" alt="Screen Shot 2021-01-28 at 17 38 17" src="https://user-images.githubusercontent.com/73803630/106126541-1165c800-6190-11eb-86c7-2042b9a095dc.png">

Stripe data received via webhook:
```
{
  ...
  "object": "refund",
  ...
  "status": "failed",
  "amount": 1557,
  ...
  "currency": "jpy",
  ...
}
```

Fixing...

kalessil commented on tests/unit/test-class-wc-payments-utils.php:
```diff
@@ -316,4 +316,22 @@ public function test_redact_array_handles_non_arrays() {
+		$original_currency = $order->get_currency();
+
+		// make sure WC_Payments::maybe_set_original_order_currency() sets _wcpay_original_order_currency.
+		$this->assertEquals( $order->get_meta( '_wcpay_original_order_currency' ), $original_currency );
```
`_wcpay_original_order_currency` -> `_wcpay_stripe_intent_currency` or something similar, if it has to remain ofcourse =)

vbelolapotkov replied:
Let's avoid Stripe here, `_wcpay_intent_currency` or `_wcpay_customer_currency` should be good enough.

shendy-a8c replied:
`_wcpay_intent_currency` it is!

vbelolapotkov commented:
@kalessil could you please finish reviews on this one? I feel I can't put enough time and attention to it.

kalessil commented:
@vbelolapotkov : gladly =)

shendy-a8c commented:
Last week, @kalessil and I had a discussion about this PR and the outcome was:
- As much as possible, get the currency from Stripe API, and
- Possibly remove the need to store 'original order currency'.

I just made the changes:
- Removed the `woocommerce_before_order_object_save` hook.
- Add new attribute, `$currency`, to `WC_Payments_API_Intention` to hold the currency value we get from the server.
- After successful intent call, save the intent currency in order's meta data to be used in places where we have to fallback to this getting currency from this order's meta data.

Questions:
1. I can only see one place to set the order's intent currency, that is [inside process_payment_for_order()](https://github.com/Automattic/woocommerce-payments/pull/1227/files#diff-bcfd303d507c33a27671935e724ee64d0df30ff2a18bb6ef4c94f28d48190124R686-R688). Do I miss anywhere else where we create intent? I searched and it seems that's it.
2. Introducing a new required parameter to `WC_Payments_API_Intention`'s constructor requires a lot of changes in unit tests and several places in the code. I was tempted to make that parameter optional and default to USD, but I decided not. What do you think?

ps: I updated the PR description.

kalessil commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -686,6 +683,9 @@ public function process_payment_for_order( $cart, $payment_information ) {
 		$order->update_meta_data( '_intent_id', $intent_id );
 		$order->update_meta_data( '_charge_id', $charge_id );
 		$order->update_meta_data( '_intention_status', $status );
+		if ( ! empty( $currency ) ) {
```
With reference to my comment above, we don't need this check and can go ahead with the call itself here.

kalessil reviewed (changes requested):
@shendy-a8c : overall looks great. I left a bunch of minor comments we need to address before we can merge the PR.

shendy-a8c commented on includes/wc-payment-api/models/class-wc-payments-api-intention.php:
```diff
@@ -56,14 +63,16 @@ class WC_Payments_API_Intention {
+	 * @param string   $currency      - The currency of the intention.
 	 */
-	public function __construct( $id, $amount, DateTime $created, $status, $charge_id, $client_secret ) {
+	public function __construct( $id, $amount, DateTime $created, $status, $charge_id, $client_secret, $currency ) {
```
Good call!

shendy-a8c commented on includes/class-wc-payments-utils.php:
```diff
@@ -119,6 +124,24 @@ public static function prepare_amount( $amount, $currency = 'USD' ) {
+	 *
+	 * @return int The interpreted amount.
+	 */
+	public static function interpret_stripe_amount( $amount, $currency = 'usd' ) {
```
Here:
- I fixed the @return phpdoc type from int to float.
- Changed the type of `$amount` from float to int.
- Set return type to be float.

Correct? Especially the `$amount`'s type.

kalessil replied:
Yes!

shendy-a8c commented:
- [x] Rebased onto trunk.
- [x] Fix changelog and readme.
- [x] Make all the proposed changes.

kalessil reviewed (approved):
LGTM, :shipit:

kalessil commented:
Marking as blocked before we have a decision if this coming in this release or the next one.

kalessil commented:
@shendy-a8c : removing the "blocked" label, go ahead with merging :tada:
